### ISSUE TITLE
New AST-based Editor API

### ISSUE THREAD

**Issue by eaburns** [eaburns/T#103]
Even after #102, I am unhappy with the `Editor` API. We should restructure it to be more like `Address`. 

`Address`es are a data structures. They can be computed on an `Editor`. They can be built either programatically using an intuitive infix notation (none of this Lisp-y prefix notation junk) or by parsing their string representation using `edit.Addr`. Anyone, even outside the edit package, can parse an address string. Once parsed, they can be serialized back to a parsable string using `.String()`. The string is just a serialized representation of the data structure. All methods use the data structure, not the string. This is beautiful.

The `Editor` API, is similar but quite different. It's similar because edits also have a nice string representation. It's different because there is no data structure for the edit commands. We either execute the commands by method call, or execute a string directly using `Editor.Edit`. It makes for strange inconsistencies.

Consider the to-be-implemented `Editor.Loop` method. Loop, or x, takes a regexp and an edit and applies the edit for each match of the regexp within an address. With the current API, it would look like this:
```go
func (ed *Editor) Loop(a Address, regexp *re1.Regexp, edit string) error
```
Now, suddenly, there is a strange mix. On one hand, the `Loop` method is part of the programatic interface for editing. On the other hand, you need to pass an edit string as one of its arguments. An ugly mix.

Another issue is parsing. Parsing is done only when executing an edit string using `Editor.Edit`. Suppose the UI wants to parse edit commands. An example of this, in Acme parlance, would be to support 2-clicking `Edit <cmd> Edit <cmd>` and expecting two edits to be performed. The UI would need to know where one edit ends and the next begins. It would need to parse the `<cmd>`s. It can't. It would either need a whole new command parser, or some type of "lite" command parser that understood the `<cmd>` delimiters. Either way, it's ugly.

Instead, edit commands should be a data structure that can be executed by an `Editor` much like `Address`es:
```go
type Edit interface {
    String() string
    do(ed *Editor, w io.Writer) // w is for any print or where output from the command.
}
func Substitute(regexp *re1.Regexp, with string) Edit
func Change(a Address, to []rune) Edit
func Loop(a Address, regexp *re1.Regexp, edit Edit) Edit
func Ed(edit string) (Edit, int, error) // Parses an edit like how Addr parses an Address.
// etc.

func (ed *Editor) Do(edit Edit, w io.Writer) error
```

Now, anyone can parse an edit using `edit.Ed`. Edits can be passed around and re-serialized into parsable strings using `.String()`. The `Editor` API is also much simpler, it consists mainly of the `Do` method.

But `Change` doesn't accept streaming input, you say? That's OK. The c (and a and i) command(s) to `Editor.Edit` don't accept streaming input either. Their input is parsed from the edit []rune. Streaming will still be available by an `Editor.Change` method. I think the final `Editor` method set will look like this:
```go
type Editor
    func NewEditor(buf *Buffer) *Editor
    func (ed *Editor) Change(a Address, r io.Reader) error // an empty or nil reader, is a delete
    func (ed *Editor) Close() error
    func (ed *Editor) Do(edit Edit, w io.Writer) error
    func (ed *Editor) Where(a Address) (from, to int64, error) // used internally too, Address.addrFrom is renamed to Address.where.
```

**Comment by mccoyst** [comment]
Sounds good in principle, but how about an example of using it?

**Comment by eaburns** [comment]
Good idea. I hadn't done this, but it is important to see what it would look like. True, that it does seem a bit less convenient than I though. I still think that I like it:

```go
// A bit verbose. I'm considering changing the API to take strings, not []runes. Still a bit unsure so let's say it takes a []rune for now.
ed.Do(Change(All, []rune{"Hello World!"}), nil) // nil means ignore any output. Change doesn't output anything.

// This fixes the issues with Editor.Substitute having too many parameters.
ed.Do(Substitute{
    Address: Dot.To(Dot.Plus(Line(10))),
    Regexp: re1.MustCompile("Whatever", re1.Options{}),  // I don't think re1.MustCompile actually exists yet.
    Global: true,
    From: 2, // Start with the second match.
}, nil)

re := re1.MustCompile("Hello, World", re1.Options{})
e := Loop(All, re, Block(Dot, []Edit{    // Block is { }. Is there a better name? Sequence? Too long?
    Change(Regexp("Hello"), []rune{"There"}),
    Move(Regexp("World"), Regexp("World").Plus(Line(1))),
    Print(Dot),
}))
ed.Do(e, w)

// Print addresses of all C-style block comments.
e, err := edit.Ed(`.+/(\/\*(([^*\/]|\n)|([^*]|\n)\/|\*([^\/]|\n))*\*\/)/p`)
if err != nil { panic(err) }
ed.Edit(e, w)
```

**Comment by mccoyst** [comment]
:+1:
